### PR TITLE
Upgrade shadow-cljs and ClojureScript

### PR DESCRIPTION
**Pull request by ilmotta**
### Summary

This PR upgrades shadow-cljs from `2.11.16` (released on 2021-02-03) to `2.19.6` (released on 2022-07-14), so ~1.5 years worth of upgrades. By upgrading shadow-cljs we can finally use the latest major Clojure version `1.11`.

### Why upgrade shadow-cljs?

- Shadow-cljs controls the ClojureScript version we can use. In order to use the latest major Clojure version (`1.11.x`) we need to upgrade shadow-cljs.

- Shadow-cljs releases new versions very frequently, and if you take a look at its [changelog](https://github.com/thheller/shadow-cljs/blob/master/CHANGELOG.md), you'll see it had tons and tons of bug fixes over the years. I hope some of them help improve the DX for certain contributors who recently reported issues with shadow-cljs. It might make things worse though, we never know.

- Clojure 1.11 brings new features, bug fixes and even performance improvements (although I think the performance mostly impacts Clojure on the JVM). See the changelog https://github.com/clojure/clojure/blob/master/changes.md#changes-to-clojure-in-version-1110

Things that can be beneficial to us, or are interesting nonetheless:

- New `:as-alias` to be used in `require`, which is like `:as` but does not require the namespace to load. This means namespaced keywords using `:as-alias` can't cause circular dependency errors. This feature would very useful if we used namespaced keywords, but we don't, so... https://github.com/clojure/clojure/blob/master/changes.md#22-as-alias-in-require
- New macros `run-test` and `run-test-var` to run single test with fixtures and report.
- New `iteration` function, useful for processing paginated data. https://www.abhinavomprakash.com/posts/clojure-iteration/
- New `update-keys` function: applies a function to every key in a map.
- New `update-vals` function: applies a function to every value in a map.

Examples for `update-vals` and `update-keys`. They should perform better than the common `reduce-kv` approach since they use a transient data structure.

```clojure
(let [m {:a 1 :b 2}]
  (update-vals m inc)) ; => {:a 2, :b 3}

(let [m {:a 1 :b 2}]
  (update-keys m name)) ; => {"a" 1, "b" 2}
```

### Why change namespaces within `__tests__` directories?

Any namespace with the word `--tests--` throws an error, like the one below. I didn't bother investigating why, so I changed the guidelines to reflect the new convention. It's probably related to the double dashes in the name :man_shrugging:

```
Namespace quo2.components.dividers.--tests--.divider-label-component-spec has a segment starting with an invalid JavaScript identifier at line 1
```

### Why can't we upgrade to the most recent shadow-cljs version?

We can't go to versions newer than `2.19.6` because we're stuck with the very old OpenJDK 8, and some new code from shadow-cljs `2.19.7+` requires at least OpenJDK 11.

### Review notes

Please, review this PR by running the app as you usually do, especially if you have access to macOS, since I don't have one at hand.

### Steps to test

Things I tested:

- Running unit tests with or without the REPL.
- Running component tests.
- Re-frisk (since I had to hard code a new version to remove a warning from the `encore` lib).
- Released APK on an Android device.
- Smoke tests in the app (clicking around).

status: ready

### CODE REVIEW

File: src/status_im2/common/bottom_sheet/view.cljs
Comment (ilmotta) [inline src/status_im2/common/bottom_sheet/view.cljs]: FYI: The newer ClojureScript version throws a runtime error when the original function signature is used. I just removed the extra arguments, since they were never used in the first place.

**Review by siddarthkay — COMMENTED** [review]
great work! 
Hopefully when the `react-native` upgrade gets merged in we have java upgraded to `openjdk11` in that branch. 
That should further unblock you to experiment with latest `shadow-cljs`.

**Review by siddarthkay — APPROVED** [review]

**Review by OmarBasem — APPROVED** [review]

**Review by yqrashawn — APPROVED** [review]

File: shadow-cljs.edn
Comment (flexsurfer) [inline shadow-cljs.edn]: https://github.com/thheller/shadow-cljs/issues/1080#issuecomment-1412641778

Comment (flexsurfer) [reply shadow-cljs.edn]: this won't work with hermes engine

**Review by alwx — APPROVED** [review]

**Review by jakubgs — APPROVED** [review]
I think you might have forgotten to run `make nix-update-clojure`.

**Review by J-Son89 — APPROVED** [review]
Nice one @ilmotta!!

File: shadow-cljs.edn
Comment (ilmotta) [reply shadow-cljs.edn]: Thanks @flexsurfer I'll check again, that's quite unfortunate. I have Hermes disabled in my environment, that's probably why it worked.

Comment (flexsurfer) [reply shadow-cljs.edn]: yeah, in latest RN its enabled by default on ios

**Comment by vkjr** [comment]
Cool! I'm glad we'll have `update-vals` 😍

**Review by vkjr — APPROVED** [review]

**Comment by OmarBasem** [comment]
> We can't go to versions newer than 2.19.6 because we're stuck with the very old OpenJDK 8

@ilmotta btw why can't we upgrade OpenJDK?

**Comment by ilmotta** [comment]
> > We can't go to versions newer than 2.19.6 because we're stuck with the very old OpenJDK 8
> 
> @ilmotta btw why can't we upgrade OpenJDK?

@OmarBasem, I don't know the answer to this question because I've never tried to upgrade. @siddarthkay said our RN version is tied to Java 8, so if we upgrade RN to let's say 0.70, then we might be able to go to Java 11.

But right now I'm having trouble upgrading shadow-cljs due to the need to support Hermes, but in order to support Hermes we need to compile CLJS to the ES5 target, but the ES5 target doesn't work with newer versions of shadow-cljs (at least in `status-mobile` repo). It's a can of worms...

Unfortunately I might just close this PR if I can't find a solution until tomorrow :/

**Comment by OmarBasem** [comment]
> Unfortunately I might just close this PR if I can't find a solution until tomorrow :/

@siddarthkay are we expecting the RN upgrade anytime soon :)

**Comment by siddarthkay** [comment]
@OmarBasem : can't say when, some serious Yak Shaving going on here : https://github.com/status-im/status-mobile/pull/15203
All I can say is that the RN Upgrade has the attention it deserves and we will get there soon :)

**Comment by ilmotta** [comment]
> @OmarBasem : can't say when, some serious Yak Shaving going on here : #15203 All I can say is that the RN Upgrade has the attention it deserves and we will get there soon :)

Yes, there's some seriously good work going on the upgrade to RN! :100: :heart: 

@OmarBasem I think it's fine if we stay on Java 8 until all the dust settles because we don't need to go to the very latest of shadow-cljs & ClojureScript. But we'll eventually get to OpenJDK 11 and beyond, I'm sure!

**Comment by ulisesmac** [comment]
> But right now I'm having trouble upgrading shadow-cljs due to the need to support Hermes, but in order to support Hermes we need to compile CLJS to the ES5 target, but the ES5 target doesn't work with newer versions of shadow-cljs (at least in `status-mobile` repo). It's a can of worms...
> 

Hey @ilmotta 
What's the problem when compiling to ES5?

Some months ago(~3), I created a small RN app using a recent version of shadow-cljs, Hermes enabled and compiled to ES5. It ran without any problems.

Also, Hermes can be disabled so that we can compile to ES6.

**Comment by ilmotta** [comment]
> > But right now I'm having trouble upgrading shadow-cljs due to the need to support Hermes, but in order to support Hermes we need to compile CLJS to the ES5 target, but the ES5 target doesn't work with newer versions of shadow-cljs (at least in `status-mobile` repo). It's a can of worms...
> 
> Hey @ilmotta What's the problem when compiling to ES5?
> 
> Some months ago(~3), I created a small RN app using a recent version of shadow-cljs, Hermes enabled and compiled to ES5. It ran without any problems.
> 
> Also, Hermes can be disabled so that we can compile to ES6.

@ulisesmac, it simply doesn't work, on my machine at least, but since we're all running through Nix, it probably won't work for you too. You can try on this branch to change the shadow-cljs setting to ES5 and it'll fail saying the G. Closure Compiler is not available. The cause of the issue came from shadow-cljs 2.15.13.

> Also, Hermes can be disabled so that we can compile to ES6.

Hermes is disabled on my machine too, but as @flexsurfer said, it's the default engine when running on macOS. If we could just say that we won't ever use Hermes then I would have been able to merge this PR already.

@flexsurfer, do we need to support Hermes at all? Hermes is already problematic for some folks, including me. When it's enabled it causes crashes on my GNU/Linux environment. What's your opinion as a macOS user?

**Comment by flexsurfer** [comment]
so, we're trying to use a tool, which allows you to set a property but it doesn't work, right, so we need to file an issue in shadow-cljs repo i guess? hermes is used in release builds, and it should be faster, and its default engine in RN, so im not sure disabling it is a good idea

**Comment by ilmotta** [comment]
> so, we're trying to use a tool, which allows you to set a property but it doesn't work, right, so we need to file an issue in shadow-cljs repo i guess? hermes is used in release builds, and it should be faster, and its default engine in RN, so im not sure disabling it is a good idea

I agree, that's the way forward. Thanks for sharing.

**Comment by ilmotta** [comment]
@ulisesmac, running the command @jakubgs kindly [reminded me](https://github.com/status-im/status-mobile/pull/15417#pullrequestreview-1349890686) did the trick, so it's working with/without Hermes. Thanks for reviewing the PR. Could you please check if this branch works on your Linux distro?

@J-Son89, @OmarBasem, @flexsurfer, could one of you please verify if the branch works with Hermes enabled. I know you guys use macOS :) I rebased against `develop` now. Everything should just work ;)

**Review by flexsurfer — APPROVED** [review]
thank you @ilmotta

**Comment by ulisesmac** [comment]
> @ulisesmac, running the command @jakubgs kindly [reminded me](https://github.com/status-im/status-mobile/pull/15417#pullrequestreview-1349890686) did the trick, so it's working with/without Hermes. Thanks for reviewing the PR. Could you please check if this branch works on your Linux distro?
> 
> @J-Son89, @OmarBasem, @flexsurfer, could one of you please verify if the branch works with Hermes enabled. I know you guys use macOS :) I rebased against `develop` now. Everything should just work ;)

I'm using Fedora, I'll try it today, sorry for the waiting.

**Comment by ilmotta** [comment]
Hey everyone, thank you for reviewing this PR. Unfortunately I'm closing this PR, since I reached a major roadblock. I'm giving as much context as I can in this comment, so in the future I can get back to this upgrade nightmare.

What happens is that the existing workaround in our codebase to support Hermes (in development) is to set the Closure Compiler feature set to `:es5`, but this causes the error below in shadow-cljs [2.15.13+](https://github.com/thheller/shadow-cljs/blob/master/CHANGELOG.md#21513---2021-11-04) (which is still a quite old version from 2021). The issue was introduced by this shadow-cljs commit https://github.com/thheller/shadow-cljs/commit/7fefe5b3f604ff11d33539e4899e160dd1ff6eb0.

```
[Tue Mar 21 2023 17:42:09.749]  ERROR    ReferenceError: Property '$jscomp' doesn't exist, js engine: hermes
[Tue Mar 21 2023 17:42:09.749]  ERROR    Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication), js engine: hermes
```

Setting the feature set to `:es6` in the `dev` target does allow us to upgrade shadow-cljs much further (up to `2.19.6`, not latest due to the limitation of the OpenJDK 8 version in use), but unfortunately it does NOT work with Hermes. Therefore, **I could not find any solution to use Hermes with shadow-cljs `2.15.13+`**. Interestingly, when compiling for `release`, the app builds and runs perfectly with or without Hermes. I've tried to use the very same release compiler options in the dev build to no avail.

I'm hopeful that once we upgrade to RN 0.70+, the new Hermes version might behave differently and give us another shot at upgrading shadow-cljs and ClojureScript. If after the RN upgrade we still can't upgrade shadow-cljs, then I'll try to report a detailed issue to the shadow-cljs maintainer.

**Comment by flexsurfer** [comment]
here https://github.com/status-im/status-mobile/issues/14776 i used latest versions of RN and shadow and it worked just fine

**Comment by ilmotta** [comment]
> here #14776 i used latest versions of RN and shadow and it worked just fine

That's great news @flexsurfer! So as soon as the RN upgrade lands in `develop` we'll be good to go!

**Comment by OmarBasem** [comment]
Hey @ilmotta, any chance to try the shadow-cljs upgrade again as we are using Java 11 now

**Comment by ilmotta** [comment]
Hi @OmarBasem. I can imagine you're still suffering from shadow-cljs issues. Let me give this issue another good try today and report back the results.

**Comment by ilmotta** [comment]
@OmarBasem, whenever you can, please test this branch with Hermes enabled and if possible, check if you can do a local release to iOS (because I can't yet test this). I've reopened the PR because I managed to upgrade to the latest Shadow CLJS and as expected, the accompanying latest ClojureScript. If you report no issues on your side I'll again request reviews from the other devs.

**Comment by OmarBasem** [comment]
Much appreciated @ilmotta. I will try working with the branch for some time and update you 👍

**Comment by OmarBasem** [comment]
Hi @ilmotta, I have tried out the branch and it seems to be working well so far 👍 

Building on Android with Hermes enabled works fine.

But I wasn't able to test local iOS release builds. For me building local ios releases fails -- even on develop, so I guess it is not something specific to this PR. I see that the iOS build in CI has succeeded.

**Comment by ilmotta** [comment]
Hey folks, I re-requested your reviews and resurrected this PR (thanks @OmarBasem) because I could upgrade Shadow CLJS and ClojureScript to their latest versions after the amazing work done in the mobile team to upgrade RN 🚀 

I haven't seen any issues in my tests with Android with and without Hermes and in dev or release builds.

**If you can, please report back if you tested on iOS and everything is good**. That way we can quickly merge this PR.

Thank you! 💟

**Review by siddarthkay — APPROVED** [review]
Great work @ilmotta!
I tested the PR build on my iPhone 11 and it works fine!

**Review by OmarBasem — APPROVED** [review]
🚀

**Review by flexsurfer — APPROVED** [review]
nice!

**Review by yqrashawn — APPROVED** [review]

**Review by alwx — APPROVED** [review]

**Review by smohamedjavid — APPROVED** [review]

**Comment by churik** [comment]
Thanks for your amazing work!
e2e failures are known but not related to PR.
Tested basic flows, LGTM!

**Devices:**
- IPhone 14 Pro (IOS 16.5.1)
- Google Pixel 7A (Android 13)

**Comment by ilmotta** [comment]
Oh @churik, thank you so much, I was going to move the PR to the e2e column but forgot in the myriad of things going on today. Great news, I want this PR to be merged real bad ;)

**Pull request merged**
